modules/apps/static/portal-osgi-web/portal-osgi-web-wab-generator/src/main/java/com/liferay/portal/osgi/web/wab/generator/internal/processor/WabProcessor.java

```diff
@@ -441,6 +441,10 @@ protected void processClass(Analyzer analyzer, String value) {
 		PackageRef packageRef = analyzer.getPackageRef(packageName);
 
 		packages.put(packageRef, new Attrs());
+
+		if (!_importPackageParameters.containsKey(packageName)) {
+			_importPackageParameters.add(packageName, _optionalAttrs);
+		}
```

Comment:
This fix is not correct. The package was already added to the set of analyzed packages. If the package was NOT contained in the JAR it would automatically be included as an import.

I think your later fix mitigates this fix by making sure the package in the custom attribute is known to the analyzer.

Reply:
Hi Ray,

During my testing, it did not appear that packages added to set of analyzed packages were being added to the Import-Package directive, even though the package was NOT container in a jar.  

Also, I was not able to find code that would do this for us. Can you direct me to the code where this logic is handled or is that something that's automatically handled by OSGI's analyzer?

Thanks!

Reply:
Yes, it's handled by the bnd analyzer.

Reply:
this line right here https://github.com/liferay/liferay-portal/blob/master/modules/apps/static/portal-osgi-web/portal-osgi-web-wab-generator/src/main/java/com/liferay/portal/osgi/web/wab/generator/internal/processor/WabProcessor.java#L638

... the magical little `*` means "... plus everything else bnd discovered that is needed".

Reply:
I think you should try with just the other part of your fix. At run-time, send it the originally failing bundle, and see what you get.

Reply:
I'll give that a try, thanks for the insight and guidance.

Reply:
Just tested with the other part of my fix and the package was included like you said.
We'll re-send a pr for this fix.

Thanks!

Reply:
:+1: you could just push updates to this one too...